jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java

```diff
@@ -0,0 +1,457 @@
+import org.eclipse.jetty.websocket.core.io.WebSocketRemoteEndpointImpl;
+
+import javax.servlet.ServletContext;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class WebSocketHandler extends HandlerWrapper
+{
+    static final Logger LOG = Log.getLogger(WebSocketHandler.class);
+
+    private static HttpField UpgradeWebSocket = new PreEncodedHttpField(HttpHeader.UPGRADE,"WebSocket");
+    private static HttpField ConnectionUpgrade = new PreEncodedHttpField(HttpHeader.CONNECTION,HttpHeader.UPGRADE.asString());
+    private static Handshaker RFC6455_Handshaker = new RFC6455Handshaker();
+
+
+    public interface Handshaker
+    {
+        boolean upgradeRequest(Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException;
+    }
+
+    interface WebSocketConnectionFactory extends ConnectionFactory
+    {
+        WebSocketCoreConnection newConnection(Connector connector, EndPoint endPoint, List<String> extensions);
+
+        @Override
+        default Connection newConnection(Connector connector, EndPoint endPoint)
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        default String getProtocol()
+        {
+            return "ws";
+        }
+
+        @Override
+        default List<String> getProtocols()
+        {
+            return Collections.singletonList(getProtocol());
+        }
+    }
+
+    interface WebSocketSessionFactory
+    {
+        // TODO why is core session abstract?
+        WebSocketCoreSession<ContainerLifeCycle,WebSocketCoreConnection,WebSocketLocalEndpoint,WebSocketRemoteEndpoint>
+        newSession(WebSocketCoreConnection connection, String subprotocol);
+    }
+
+
+    public WebSocketHandler()
+    {
+    }
+
+    @Override
+    public void handle(String target, Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException
+    {
+        Handshaker handshaker = getHandshaker(baseRequest);
+        if (LOG.isDebugEnabled())
+            LOG.debug("handle {} handshaker={}",baseRequest,handshaker);
+        if (handshaker!=null && handshaker.upgradeRequest(baseRequest, request, response))
+            return;
+        super.handle(target,baseRequest,request,response);
+    }
+
+
+    protected Handshaker getHandshaker(Request baseRequest)
+    {
+        // TODO this can eventually be made pluggable
+        HttpField version = baseRequest.getHttpFields().getField(HttpHeader.SEC_WEBSOCKET_VERSION);
+        if (version !=null && version.getIntValue()==RFC6455Handshaker.VERSION)
+            return RFC6455_Handshaker;
+        return null;
+    }
+
+
+    private final static class RFC6455Handshaker implements Handshaker
```

Comment:
This is what websocket-servlet does

Reply:
OK - missed that, but a whole package for one class could be bit heavy.